python/tank/util/authentication.py

```diff
@@ -0,0 +1,314 @@
+
+        login = config.get("LoginInfo", "login", raw=True)
+        session_token = config.get("LoginInfo", "session_token", raw=True)
+
+        if not login or not session_token:
+            logger.debug("Incomplete settings (login:%s, session_token:%s)", login, session_token)
+            return None
+
+        return {"login": login, "session_token": session_token}
+    except Exception:
+        logger.exception("Exception thrown while loading cached session info.")
+        return None
+
+
+def _validate_session_token(host, session_token, http_proxy):
+    """
+    Validates the session token by attempting to an authenticated request on the site.
+    :param session_token: Session token to use to connect to the host.
+    :param host: Host for that session
+    :param http_proxy: http_proxy to use to connect. If no proxy is required, provide None or an empty string.
+    """
+    # Connect to the site
+    logger.debug("Creating shotgun instance")
+    sg = _shotgun_instance_factory(
+        host,
+        session_token=session_token,
+        http_proxy=http_proxy
+    )
+    try:
+        sg.find_one("HumanUser", [])
+        return sg
+    except AuthenticationFault, e:
+        # Session was expired.
+        logger.exception(e)
+        return None
+
+
+def cache_session_data(host, login, session_token):
+    """
+    Caches the session data for a site and a user.
+    :param host: Site we want to cache a session for.
+    :param login: User we want to cache a session for.
+    :param session_token: Session token we want to cache.
+    """
+    # Retrieve the cached info file location from the host
+    info_path = _get_cached_login_info_location(host)
+
+    # make sure the info_dir exists!
+    info_dir, info_file = os.path.split(info_path)
+    path.ensure_path_exists(info_dir)
+
+    logger.debug("Caching login info at %s...", info_path)
+    # Create a document with the following format:
+    # [LoginInfo]
+    # login=username
+    # session_token=some_unique_id
+    config = SafeConfigParser()
+    config.add_section("LoginInfo")
+    config.set("LoginInfo", "login", login)
+    config.set("LoginInfo", "session_token", session_token)
+    # Write it to disk.
+    with open(info_path, "w") as configfile:
+        config.write(configfile)
+    logger.debug("Cached!")
+
+
+def _delete_session_data():
+    """
+    Clears the session cache for a given site.
+    """
+    host = shotgun.get_associated_sg_base_url()
+    logger.debug("Clearing session cached on disk.")
+    try:
+        info_path = _get_cached_login_info_location(host)
+        if os.path.exists(info_path):
+            logger.debug("Session file found.")
+            os.remove(info_path)
+            logger.debug("Session cleared.")
+        else:
+            logger.debug("Session file not found: %s", info_path)
+    except:
+        logger.exception("Couldn't delete the site cache file")
+
+
+def generate_session_token(hostname, login, password, http_proxy):
```

Comment:
Would it make sense if this also returned the sg instance?

What's the current flow around login? The authentication code calls this to attempt to log in and generate a token, then if successful, this token is used to generate a shotgun connection? Wondering if it's worth (from a performance POV) to try to use the sg connection used here directly? Or would you foresee any security problems with that approach?